scalgrafanalib/core.py

```diff
@@ -99,19 +131,96 @@ def to_json_data(self):
         }
 
 
+@attr.s
+class StateMapping:
+    """
+    Represents a state mapping for StateTimeline panels with color and text
+    """
+
+    value: Union[int, str, None] = attr.ib()
+    text: str = attr.ib()
+    color: Optional[str] = attr.ib(default=None)
+    index: int = attr.ib(default=0)
+
+    def to_json_data(self) -> Dict[str, Any]:
+        result = {
+            "index": self.index,
+            "text": self.text,
+        }
+        if self.color:
+            result["color"] = self.color
+        return result
+
+
 @attr.s
 class StateTimeline(core.StateTimeline):
-    """StateTimeline: Allow settings minValue and maxValue"""
+    """StateTimeline: Allow settings minValue, maxValue, and color mappings for MongoDB states"""
 
     minValue = attr.ib(default=None)  # pylint: disable=invalid-name
     maxValue = attr.ib(default=None)  # pylint: disable=invalid-name
+    colorMode: str = attr.ib(default="palette-classic")  # pylint: disable=invalid-name
+    mappings: List[Union[StateMapping, Dict[str, Any]]] = attr.ib(
+        default=None
+    )  # pylint: disable=invalid-name
 
     def to_json_data(self) -> Json:
         json = super().to_json_data()
         if self.minValue:
             json["fieldConfig"]["defaults"]["min"] = self.minValue
         if self.maxValue:
             json["fieldConfig"]["defaults"]["max"] = self.maxValue
+
+        # Set color mode
+        if "fieldConfig" not in json:
+            json["fieldConfig"] = {}
+        if "defaults" not in json["fieldConfig"]:
+            json["fieldConfig"]["defaults"] = {}
+        if "color" not in json["fieldConfig"]["defaults"]:
+            json["fieldConfig"]["defaults"]["color"] = {}
+        json["fieldConfig"]["defaults"]["color"]["mode"] = self.colorMode
+
+        # Add mappings if provided
+        if self.mappings:
+            if "mappings" not in json["fieldConfig"]["defaults"]:
+                json["fieldConfig"]["defaults"]["mappings"] = []
+
+            # Create value mapping options
+            value_options = {}
+            for mapping in self.mappings:
+                # Handle both StateMapping objects and dictionaries
+                if isinstance(mapping, StateMapping):
+                    # StateMapping object
+                    key = str(mapping.value) if mapping.value is not None else "null"
+                    value_options[key] = mapping.to_json_data()
+                else:
```

Comment:
do we really need to handle that case? would be simpler to just "override" the `mapping` field, and expect (in scalgrafanalib) a StateMapping

Also it seems to me that just putting an object with `to_json_data` instead of a json object should be enough, thanks to [the magic in grafanalib](https://github.com/weaveworks/grafanalib/blob/v0.7.1/grafanalib/core.py#L1971) -- like it happens for panels in Dashboard or Row -- though I could not locate the place where the recursion through lists/arrays is handled : if you did not try it, can you check that passing a list of `StateMapping` in the "existing" grafalnalib `mapping` field does not yield the expected result? (would avoid lots of boiler plate here...)

Reply:
I ended up using `StatValueMappings` from grafanalib, that removed this logic.